client/src/controllers/RevealController.ts

```diff
@@ -0,0 +1,311 @@
+  declare readonly hasCloseIconClass: string;
+  declare readonly hasContentTarget: boolean;
+  declare readonly hasOpenIconClass: string;
+  declare readonly hasToggleTarget: boolean;
+  declare readonly initialClasses: string[];
+  declare readonly openedClasses: string[];
+  declare readonly openedContentClasses: string[];
+  declare readonly openIconClass: string;
+  declare readonly peekTargetValue: string;
+  declare readonly toggleTarget: HTMLButtonElement;
+  declare readonly toggleTargets: HTMLButtonElement[];
+
+  cleanUpPeekListener?: () => void;
+
+  /**
+   * Ensure that targets are correctly set up, as this runs
+   * before the initial call to `closedValueChanged`.
+   * If there is no content element set, try to set one
+   * via `aria-controls` on the toggle, this also ensures
+   * that the toggleTarget has been set before the controller
+   * connects.
+   */
+  initialize() {
+    const toggleTarget = this.toggleTarget;
+
+    if (!this.hasContentTarget) {
+      const ariaControlsSelectors = (
+        toggleTarget.getAttribute('aria-controls') || ''
+      )
+        .split(' ')
+        .filter(Boolean)
+        .map((id) => `#${id}`)
+        .join(',');
+
+      const contentElements = ariaControlsSelectors
+        ? document.querySelectorAll(ariaControlsSelectors)
+        : [];
+
+      if (!contentElements.length) {
+        const error = `data-${this.identifier}-target="content" must be set or aria-controls must target a valid element.`;
+        throw new Error(error);
+      }
+
+      contentElements.forEach((content) => {
+        content.setAttribute(`data-${this.identifier}-target`, 'content');
+      });
+    }
+
+    // Only attempt to set to closed if not already closed via initial value
+    const isClosed = this.closedValue;
+    if (isClosed) return;
+
+    const ariaExpanded = toggleTarget.getAttribute('aria-expanded');
+    const isClassNameClosed = this.hasClosedClass
+      ? this.element.classList.contains(this.closedClass)
+      : false;
+
+    if (isClassNameClosed && (!ariaExpanded || ariaExpanded === 'false')) {
+      this.closedValue = true;
+    } else if (ariaExpanded === 'false') {
+      this.closedValue = true;
+    }
+  }
+
+  connect() {
+    // Adopt any classes we can from the toggle's icon if it has one.
+
+    const [toggle] = this.toggles;
+    const icon = toggle.querySelector<HTMLSpanElement>('.icon');
+    const hasCloseIconClass = this.hasCloseIconClass;
+    const hasOpenIconClass = this.hasOpenIconClass;
+
+    const iconClass =
+      icon &&
+      (!hasCloseIconClass || !hasOpenIconClass) &&
+      [...icon.classList].find((className) => className.startsWith('icon-'));
+
+    if (iconClass) {
+      if (!hasCloseIconClass) {
+        if (iconClass) {
+          this.element.setAttribute(
+            `data-${this.identifier}-close-icon-class`,
+            iconClass,
+          );
+        }
+      } else if (!hasOpenIconClass) {
```

Comment:
Again, I don't know why we need to be "smart" about this. I think that instead of trying to figure out and resolve values/classes that are unset in the HTML, it's better if we agree on what the expected attributes for this controller are, and make it so that the HTML templates are consistently configured. At least for now, it's much easier to understand that way, as the code doesn't have to deal with cases like "what if attribute X isn't set?".

There are cases where optional attributes make sense and it's pretty clear to give sensible defaults, but at this point most of these cases still feel very hypothetical and make the code harder to review.

In other words: if you're configuring the HTML in a way that we don't currently use within Wagtail, it's going to result in an error (like the missing `content` target) or undefined behaviour. If we do find use cases where it's possible to reuse the code with a slightly different attribute configuration, *then* we start handling such variations and making the controller code more robust.

If we do it too early, it's premature abstraction. We don't yet know what the other use cases would look like, so the abstraction/handling we put in place may or may not be necessary in the future, and if it does, it may not retain its shape as-is. Until then, it's hard to read the code without imagining what exactly the case is here.

Reply:
Agreed, this can be removed and added back later when needed for the panels or maybe we just change the approach.

Reply:
This has been removed.